determined-ai/determined
# fix: reject reconnecting agents with different resource pool configuration

## Ticket

[CM-186](https://hpe-aiatscale.atlassian.net/browse/CM-186)


## Description

Resource pool change does not apply if agent restarts too quickly. Resource Pool change on reconnect, discard the corresponding agent state and tell the agent to shutdown (and restart) as a way to not overly complicate the restore process and keep sanity.

**Error thrown by master:**
```
[ERRO] change in agent resource pool was detected    component=agent error="resource pool has changed: pool1 -> pool2" id=agent1 resource-pool=pool1
[ERRO] agent crashed                                 address=127.0.0.1 component=agent error="resource pool has changed: pool1 -> pool2" id=agent1 resource-pool=pool1 started=true
```
**Error thrown by agent before it dies:**
```[FATA] agent is past reconnect period, it must restart ```

**Previously, the steps were as follows:**

1. The master starts and creates resource pools.
2. The agent starts and connects to the master.
3. The master assigns the agent to the resource pool specified in the configuration.
4. The configuration is updated.
5. The agent is stopped, and during this time, the master attempts to reconnect/restore the agent for a few seconds.
6. If the agent restarts while the master is still trying to reconnect, the master does not reload the configuration and continues using the old resource pool.
7. If the agent does not restart within those few seconds, the master drains/removes the agent. If the agent starts afterward, the master reads the updated configuration and attempts to connect the agent to the new resource pool.

**After my code changes, the steps are now:**

1. The master starts and creates resource pools.
2. The agent starts and connects to the master.
3. The master assigns the agent to the resource pool specified in the configuration.
4. The configuration is updated.
5. The agent is stopped, and during this time, the master attempts to reconnect/restore the agent for a few seconds.
6. **If the agent tries to restart while the master is still attempting to reconnect, the master will detect the change in the resource pool, throw an error, and remove the agent, prompting a restart of the agent.**
7. If the agent does not restart within those few seconds, the master drains/removes the agent. If the agent starts afterward, the master reads the updated configuration and attempts to connect the agent to the new resource pool.


## Test Plan


- [X] Automated E2E test 

**To test manually:**

1. Update resource pool config and have agent reconnect back to master
2. Agent state will be wiped and it will be told it needs to restart
3. After another reconnection agent and master should work ok

**Commands to make/run/test manually:**
Build only master and agent to get their binaries:
`make -C master build`
`make -C agent build `
Run the docker to create a db: <you can get this command by ‘Copy the docker run’ from already running docker container> 
`% docker run --hostname=d07083bced2f --mac-address=02:42:ac:11:00:02 --env=POSTGRES_DB=determined --env=POSTGRES_PASSWORD=postgres --env=PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/lib/postgresql/10/bin --env=GOSU_VERSION=1.12 --env=LANG=en_US.utf8 --env=PG_MAJOR=10 --env=PG_VERSION=10.14-1.pgdg90+1 --env=PGDATA=/var/lib/postgresql/data --volume=/Users/shreya/.postgres:/var/lib/postgresql/data --volume=/var/lib/postgresql/data -p 5432:5432 --restart=no --runtime=runc -d postgres:10.14 `
Run the master.yaml: 
`% master/build/determined-master --config-file determined/.circleci/devcluster/master.yaml`
Run the agent.yaml:
`% agent/build/determined-agent --config-file determined/.circleci/devcluster/agent.yaml`
Run E2E test:
`% pytest e2e_tests/tests/cluster/test_master_restart.py -k "test_agent_resource_pool_change" --log-cli-level=info --user-password=<INITIAL_USER_PASSWORD>`


## Checklist

  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.




[CM-186]: https://hpe-aiatscale.atlassian.net/browse/CM-186?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
